### PR TITLE
Dynamic sparse feature allocation for flexible incremental training

### PR DESCRIPTION
**Pull request by jupyterjazz**
PR for issue #8232 
Combined version of the following PRs: #8771 , #8795 , #8796 
**Proposed changes**:
- New sparse features will be added dynamically during incremental training instead of creating 
  an extra buffer in advance
- No need to specify `additional_vocabulary_size` for `CountVectorsFeaturizer` or
`number_additional_patterns` for `RegexFeaturizer`.
- Load and store sparse feature sizes with the model
- `DenseForSparse` layers will be dynamically adjusted according to old and current sparse feature sizes 

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: tests/nlu/selectors/test_response_selector.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
+from pathlib import Path
```
Comment (dakshvar22) [inline tests/nlu/selectors/test_response_selector.py:1]: Sorry, i didn't see this earlier, but there is already a file named `test_selectors.py` where all of the content of this file can go in. We don't need a separate file.

File: docs/docs/migration-guide.mdx [hunk 2]
```diff
@@ -10,6 +10,39 @@ description: |
 This page contains information about changes between major versions and
 how you can migrate from one version to another.
 
+## Rasa 2.7 to 2.8
+
+:::caution
+This is a release **breaking backwards compatibility of machine learning models**.
+I don't think we need this, but will still leave here to discuss.
+:::
```
[the remark this answers is not in the document]
Comment (jupyterjazz) [reply docs/docs/migration-guide.mdx:18]: oh, sure. missed that. 
also featurizer vocabularies will have extra buffers.

File: changelog/8232.improvement.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+Space for new sparse features are created dynamically inside the downstream machine learning 
+models - `DIETClassifier`, `ResponseSelector`. In other words, no extra buffer is created in 
+advance for additional vocabulary items and space will be dynamically allocated for them inside the model.
+
+This means there's no need to specify `additional_vocabulary_size` for [`CountVectorsFeaturizer`](./components.mdx#countvectorsfeaturizer) or
+`number_additional_patterns` for [`RegexFeaturizer`](./components.mdx#regexfeaturizer). These parameters are now deprecated.
+
+Also, all custom layers specifically built for machine learning models - `RasaSequenceLayer`, `RasaFeatureCombiningLayer` 
+and `ConcatenateSparseDenseFeatures` now inherit from `RasaCustomLayer` so that they support flexible incremental training out of the box.
```
Comment (dakshvar22) [inline changelog/8232.improvement.md:11]: Add this after the example. Breaks the flow otherwise.

**Review by dakshvar22 — COMMENTED** [review]
Looking great 🚀 Since this is the final PR to be merged into main, please trigger model regression tests on all configs and datasets. Post that, we can merge it 🎉

**Comment by jupyterjazz** [comment]
/modeltest
```yml
include:
  - dataset: ["all"]
    config: ["all"]
```

**Review by dakshvar22 — APPROVED** [review]
Amazing! Congrats on your first merge 🎉 🥳

**Pull request merged**